File: wakakusa/demo/control.rb
# Ordinary Ruby inside a view: `if`, `unless`, a ternary, a loop, and a
# method that answers part of the screen. Nothing here is a special
# form — the block is Ruby, and each element joins the container that
# is open.
#
# The one shape that is refused is a block written inside a loop's
# block. `line` below is what to write instead, and `wakakusa check`
# says so with the line.
require "wakakusa"

class Control
  def initialize
    @items = ["milk", "eggs", "rice"]
    @picked = -1
    @show_hint = true
    @tab = 0
  end

  def hint
    text "pick one", size: 12.0, color: "#8a8f98"
  end

  def line(name, i)
    row(spacing: 8.0) {
      text(i == @picked ? "▸ #{name}" : "  #{name}")
      button("pick #{i}") { @picked = i }
    }
  end

  def tab_button(n)
    button("tab #{n}") { @tab = n }
  end

  def view
    column(spacing: 10.0, padding: 14.0) {
      text "control flow", size: 18.0, bold: true

      if @show_hint
        hint
      else
        text "hidden", size: 12.0
      end

      @items.each_with_index { |name, i| line(name, i) }

      unless @picked < 0
        text "picked #{@items[@picked]}", color: @picked.even? ? "accent" : "#f38ba8"
      end

      row(spacing: 6.0) {
        button("hint") { @show_hint = !@show_hint }
        3.times { |n| tab_button(n) }
      }
      text "tab #{@tab}"
    }
  end
end

run(Control.new, title: "control")

File: rakugan/demo/control.pl
# Ordinary Perl inside a view: `if`, `unless`, a conditional
# expression, a loop, and a method that answers part of the screen.
# The parts go into a list, and the list is what the container holds.
use Rakugan;

class Control {
    use Rakugan;
    field @items     = ("milk", "eggs", "rice");
    field $picked    = -1;
    field $show_hint = true;
    field $tab       = 0;

    method hint {
        return text("pick one", size => 12, color => "#8a8f98");
    }

    method line :Sig(Str, Int) ($name, $i) {
        return row(
            text($i == $picked ? "▸ $name" : "  $name"),
            button("pick $i", on_click => sub { $picked = $i }),
            spacing => 8,
        );
    }

    method tab_button :Sig(Int) ($n) {
        return button("tab $n", on_click => sub { $tab = $n });
    }

    method view {
        my @cells = (text("control flow", size => 18, bold => true));
        if ($show_hint) {
            push @cells, $self->hint;
        } else {
            push @cells, text("hidden", size => 12);
        }
        for my $i (0 .. $#items) {
            push @cells, $self->line($items[$i], $i);
        }
        push @cells, text("picked $items[$picked]", color => $picked % 2 == 0 ? "accent" : "#f38ba8")
            unless $picked < 0;
        push @cells, row(
            button("hint", on_click => sub { $show_hint = !$show_hint }),
            (map { $self->tab_button($_) } 0 .. 2),
            spacing => 6,
        );
        push @cells, text("tab $tab");
        return column(@cells, spacing => 10, padding => 14);
    }
}

run(Control->new, title => "control");
